A55-xds-stateful-session-affinity.md

```diff
@@ -0,0 +1,516 @@
+  enum HealthStatus {
+    UNKNOWN = 0;
+    HEALTHY = 1;
+    DRAINING = 3;
+  }
+
+  // valid health status for hosts that are considered when using
+  // xds_override_host_experimental policy.
+  // Default is [UNKNOWN, HEALTHY]
+  repeated HealthStatus override_host_status = 1;
+
+  repeated LoadBalancingConfig child_policy = 2;
+}
+```
+
+The `xds_cluster_resolver_experimental` config will be modified so that the
+`DiscoveryMechanism` inner message contains an additional
+`override_host_status` field of type `repeated HealthStatus`.
+
+`cds_experimental` policy will copy the
+[`common_lb_config.override_host_status`][or-host-status] value it gets
+in the CDS update from the `XdsClient` to the appropriate `DiscoveryMechanism`
+entry's `override_host_status` field in the config of the
+`xds_cluster_resolver_experimental` policy. If the field is unset,
+we treat it the same as if it was explicitly set to the default set which is
+[`UNKNOWN`, `HEALTHY`].
+
+The `xds_cluster_resolver_experimental` LB policy will copy the
+`override_host_status` value from its config's appropriate `DiscoveryMechanism`
+entry into the config for the `xds_override_host_experimental` policy. The
+diagram below shows how the configuration is passed down the hierarchy all the
+way from `cds_experimental` to `xds_override_host_experimental`.
+
+One of the existing policies (`xds_wrr_locality_experimental`,
+or `ring_hash_experimental`) is created as the child policy of
+`xds_override_host_experimental` as shown in the diagram below.
+
+![LB Policy Hierarchy Diagram](A55_graphics/lb-hierarchy.png)
+
+To propagate [`common_lb_config.override_host_status`][or-host-status] the
+following changes are required:
+
+* The [XdsClient][grpc-client-arch] will parse the
+  [common_lb_config.override_host_status][or-host-status] field from the
+  `Cluster` resource and copy the value to a new `override_host_status` field
+  in the CDS update that it sends to the watchers.
+
+* `cds_experimental` policy (which receives the CDS update) will copy the
+  value of `override_host_status` into the appropriate `DiscoveryMechanism`
+  entry's `override_host_status` field in the config of the
+  `xds_cluster_resolver_experimental` policy.
+
+* When the `xds_cluster_resolver_experimental` policy generates the config for
+  each priority, it will configure the `xds_override_host_experimental` policy
+  underneath the `xds_cluster_impl_experimental` policy. The generated config
+  for the `xds_override_host_experimental policy` will include the
+  `override_host_status` field from the `DiscoveryMechanism` associated with
+  that priority.
+
+### Notes
+
+Note that if stateful session affinity and retries (gRFC [A44][]) are both
+configured, all retry attempts will be routed to the same endpoint.
+
+### Temporary environment variable protection
+
+During initial development, this feature will be enabled by the
+`GRPC_EXPERIMENTAL_XDS_ENABLE_HOST_OVERRIDE` environment variable. This
```

Comment:
There are a bunch of places in this design where we use the words "host" and "override", but we're not consistent about the order -- in some places, we use "host" first, and in other places, we use "override" first.  @eugeneo has suggested that we should pick one order and be consistent about it throughout, and I agree.

I think this affects at least the following places in this design (I might have missed something):

- the name of the LB policy
- the name of the pick argument
- the name of the env var
- the name of the health status field in the CDS update struct delivered by the XdsClient to the watcher
- the name of the health status field in the `xds_cluster_resolver` policy config
- the name of the health status field in the `xds_override_host` policy config

Logically, I think it makes sense to use "host_override" instead of "override_host", because the "host" part is the noun, and the "override" part says what we're doing with that noun.  But it's worth noting that the CDS `common_lb_config.override_host_status` field uses "override" first, so that's one place where we would be inconsistent.

Ultimately, I don't really care which one we use, but let's pick one and be consistent.

Reply:
I agree that we should be internally consistent about names and I will do that. However [Envoy](https://www.envoyproxy.io/docs/envoy/v1.24.1/intro/arch_overview/upstream/load_balancing/override_host.html?highlight=override) and [xDS](https://github.com/envoyproxy/envoy/blob/v1.24.1/api/envoy/config/cluster/v3/cluster.proto#L672) use the same concept/terminology and I am assuming we don't care about consistency with them?

Reply:
I don't really care about naming consistency with Envoy's implementation.

As I mentioned above, we can't change the name of the `common_lb_config.override_host_status` field, so that is a point in favor of using "override_host", even though "host_override" seems clearer to me.  But I'd also be fine with using "host_override" internally, even though the xDS field name has the opposite order.

Ultimately, there are reasons for going with either order, and neither set of reasons is necessarily super compelling.  I think the important thing here is that we be consistent internally, so either way is fine.

Reply:
I was starting to make the change but then we cannot rename `common_lb_config.override_host_status`  and then reading the text with `...override_host_...` in some places and `...host_override...` in others becomes very confusing so I think it is better to leave it as it is.

Reply:
If you want to stick with "override_host" as the preferred order, that's fine.  But then let's change the env var name accordingly.

Reply:
done